hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/StorageVolume.java

```diff
@@ -509,21 +545,110 @@ public void failVolume() {
+              file);
+        }
+      });
+    } catch (IOException ex) {
+      LOG.warn("Failed to list contents of volume health check directory {} " +
+          "for deleting.", diskCheckDir);
+    }
   }
 
   /**
    * Run a check on the current volume to determine if it is healthy.
    * @param unused context for the check, ignored.
    * @return result of checking the volume.
+   * @throws InterruptedException if there was an error during the volume
+   *    check because the thread was interrupted.
    * @throws Exception if an exception was encountered while running
-   *            the volume check.
+   *    the volume check and the thread was not interrupted.
    */
   @Override
-  public VolumeCheckResult check(@Nullable Boolean unused) throws Exception {
-    if (!storageDir.exists()) {
+  public synchronized VolumeCheckResult check(@Nullable Boolean unused)
+      throws Exception {
+    boolean directoryChecksPassed =
+        DiskCheckUtil.checkExistence(storageDir) &&
+        DiskCheckUtil.checkPermissions(storageDir);
+    // If the directory is not present or has incorrect permissions, fail the
+    // volume immediately. This is not an intermittent error.
+    if (!directoryChecksPassed) {
+      if (Thread.currentThread().isInterrupted()) {
+        throw new InterruptedException("Directory check of volume " + this +
+            " interrupted.");
+      }
+      return VolumeCheckResult.FAILED;
+    }
+
+    // If IO test count is set to 0, IO tests for disk health are disabled.
+    if (ioTestCount == 0) {
+      return VolumeCheckResult.HEALTHY;
+    }
+
+    // Since IO errors may be intermittent, volume remains healthy until the
+    // threshold of failures is crossed.
+    boolean diskChecksPassed = DiskCheckUtil.checkReadWrite(storageDir,
+        diskCheckDir, healthCheckFileSize);
+    currentIOTestCount.incrementAndGet();
+    if (!diskChecksPassed) {
+      if (Thread.currentThread().isInterrupted()) {
+        throw new InterruptedException("IO check of volume " + this +
+            " interrupted.");
+      }
+      currentIOFailureCount.incrementAndGet();
+    }
+
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Volume {} ran IO test {} of {}. Seen {} of {} tolerated IO" +
+              "failures.", this, currentIOTestCount, ioTestCount,
+          currentIOFailureCount, ioFailureTolerance);
+    }
+
+    // If the failure threshold has been crossed, fail the volume without
+    // further scans.
+    // Once the volume is failed, it will not be checked anymore.
+    // The failure counts can be left as is.
+    if (currentIOFailureCount.get() > ioFailureTolerance) {
       return VolumeCheckResult.FAILED;
     }
-    DiskChecker.checkDir(storageDir);
+
+    // If we have completed the required number of volume checks without too
+    // many IO failures, reset the counters for future iterations.
+    if (currentIOTestCount.get() == ioTestCount) {
```

Comment:
currently tolerance is checked with-in group of ioTestCount, but not across groups with consecutive failure (like sliding window of ioTestCount size).
So will the logic be need check within absolute group size as currently Or Planning to go with sliding window ?

Reply:
I considered a sliding window as well, but did it this way since maintaining the counters was simpler. Since ioTestCount should be small, a sliding window with a queue of max size ioTestCount for each volume should be fine. I will switch to this implementation.